internal/fwschemadata/data_set_at_path.go

```diff
@@ -28,9 +28,52 @@ import (
 // Lists can only have the next element added according to the current length.
 func (d *Data) SetAtPath(ctx context.Context, path path.Path, val interface{}) diag.Diagnostics {
 	var diags diag.Diagnostics
-
 	ctx = logging.FrameworkWithAttributePath(ctx, path.String())
 
+	if v, ok := val.(tftypes.Value); ok {
+		atPath, atPathDiags := d.Schema.AttributeAtPath(ctx, path)
+
+		diags.Append(atPathDiags...)
+
+		if diags.HasError() {
+			return diags
+		}
+
+		attrType := atPath.GetType().TerraformType(ctx)
+
+		if !attrType.Equal(v.Type()) {
+			diags.AddAttributeError(
+				path,
+				d.Description.Title()+" Write Error",
+				"An unexpected error was encountered trying to write the "+d.Description.String()+". This is always an error in the provider. Please report the following to the provider developer:\n\n"+
+					fmt.Sprintf("Error: Type of provided value does not match type of %s", path.String()),
```

Comment:
Same comment as `data_set`, we should consider adding the expected + actual types to the diagnostic